### ISSUE TITLE
Shared intermediates in per-column transition / energy / gradient evaluation

### ISSUE THREAD

**Issue by fliingelephant** [fliingelephant/VMC#94]
## Summary

During per-column sweeps in both `estimate` and `transition`, several operations share the same prefix contraction. There are two levels of redundancy:

1. **F2 (CSE-level):** multiple operations begin with the same pairwise contraction, so XLA CSE may be able to eliminate it automatically.
2. **F4/F5 (Structural):** a larger monolithic contraction already contains a reusable intermediate, but that intermediate is rebuilt later by a separate helper call. This requires code restructuring; XLA will not discover it for us.

This issue was originally about the backward-pass `estimate` path. The same structural pattern also appears in `transition`, including BlockadePEPS. So the scope is not just standard PEPS local-energy evaluation.

---

## F2: Shared first contraction across three operations in `estimate`

At each column in the `dr=1` backward pass, three operations execute back-to-back:

```python
# (A) Gradient
env_grad = _compute_single_gradient(left_env, right_envs[col], top[col], bot[col])
#   einsum: ace,aub,evf,bdf -> uvcd
#   optimal path first step: left_env(ace) × top(aub) -> (c,u,e,b)

# (B) Horizontal term eval (if terms exist at this column)
val = _eval_term(horizontal_term, envs, ...)
#   einsum: ace,aub,edf,pudcr,qvwrx,bvg,fwi,gxi -> pq
#   optimal path first step: left_env(ace) × top[col](aub) -> same (c,u,e,b)

# (C) Left-env update
left_env = _update_left_env_1row(left_env, top[col], mpo[col], bot[col])
#   einsum: ace,aub,cduv,evf -> bdf
#   optimal path first step: left_env(ace) × top(aub) -> same (c,u,e,b)
```

All three share the same first pairwise contraction `left_env(a,c,e) × top(a,u,b)`, contracting over `a = Dc`.

Since all three are in the same traced function with identical input arrays, **XLA CSE may eliminate the redundancy**. No code change is necessarily required here, but it is worth verifying with profiling that CSE actually triggers.

---

## F4: Decomposing the horizontal 8-tensor einsum in `estimate`

The horizontal two-site `_eval_term` computes amplitudes for all physical index pairs via a monolithic 8-tensor contraction:

```python
amps = jnp.einsum(
    "ace,aub,edf,pudcr,qvwrx,bvg,fwi,gxi->pq",
    left_env, top[col], bot[col],
    tensor[col], tensor[col+1],
    top[col+1], bot[col+1], right_envs[col+1],
)
```

This can be split at the bond between columns `col` and `col+1` into:

1. a left half that keeps physical index `p` open,
2. a right half that keeps physical index `q` open,
3. a final join over the shared bond indices.

This is a structural optimization, not a path-tuning issue. In particular, the left half for the physical slice `p = spins[row, col]` is exactly the same object as the left-env update already computed at this column. One can therefore avoid rebuilding part of the horizontal evaluation by making that prefix explicit.

This decomposition is outside `opt_einsum`'s search space because it relies on slicing along the small physical index rather than only reordering pairwise contractions.

---

## F5: The same structural redundancy occurs in `transition`

The same pattern appears in Metropolis transition sweeps.

### Standard PEPS

In the standard transition path, the proposed amplitude is computed from

```python
(left_env, top[col], proposed_mpo, bottom[col], right_envs[col])
```

and then, after accept/reject, the code advances the sweep with

```python
left_env = _update_left_env_1row(left_env, top[col], accepted_mpo, bottom[col])
```

But the proposal amplitude contraction already contains that next left-prefix environment as an intermediate. If the move is accepted, the accepted prefix is therefore rebuilt.

### BlockadePEPS, 1-row sweep

For blockade single-row sweeping, the proposal is a 2-column window contraction. Algebraically,

```python
_contract_1row_2col(left_env, ..., mpo_c, mpo_c1, ..., right_env[c+1])
```

factorizes as

```python
prefix = _update_left_env_1row(left_env, top[c], mpo_c, bottom[c])
amp = _contract_1row_1col(prefix, top[c+1], mpo_c1, bottom[c+1], right_env[c+1])
```

So if the move is accepted, the later `_update_left_env_1row(...)` call rebuilds the same prefix.

### BlockadePEPS, 2-row sweep

For blockade 2-row sweeping, the same idea holds with the larger local window:

```python
_contract_2row_2col(left_env, top, mpo0_c, mpo1_c, mpo0_c1, mpo1_c1, bottom, right_env[c+1])
```

factorizes as

```python
prefix = _update_left_env_2row(left_env, top[c], mpo0_c, mpo1_c, bottom[c])
amp = _contract_2row_1col(prefix, top[c+1], mpo0_c1, mpo1_c1, bottom[c+1], right_env[c+1])
```

Again, if the move is accepted, the accepted prefix is recomputed when advancing the sweep.

### Cost relevance

This transition-side redundancy is paid only on **accepted** proposals, so the expected win scales with acceptance rate. It is therefore smaller than a worst-case static flop count might suggest, but it is still a real structural inefficiency.

---

## Scope note

This issue therefore covers a broader pattern:

- `estimate` path in standard PEPS,
- `transition` path in standard PEPS,
- `transition` path in BlockadePEPS.

I have not yet audited GI-PEPS carefully enough to claim the same there.

---

## Possible clean implementation direction

The cleanest fix is to expose the shared prefix explicitly, rather than introducing new wrappers or abstractions.

For any proposal helper, return both:

```python
(prefix_env, proposed_amp, updated_local_window)
```

Then on accept:

- reuse `prefix_env` as the next carry,
- reuse the updated local MPO/window tensors,
- avoid calling `_update_left_env_*` again.

This keeps the sweep logic unchanged while removing the structural duplication.
